src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -1320,6 +1320,131 @@ static void Main()
         }
 
         [Fact]
+        [WorkItem(18554, "https://github.com/dotnet/roslyn/issues/18554")]
+        public void AssigningIntoIndexers()
+        {
+            string source = @"
+using System;
+class C
+{
+    int field;
+    ref int this[int x, int y, int z, int opt = 1]
+    {
+        get
+        {
+            Console.WriteLine($""this[x: {x}, y: {y}, z: {z}].get"");
+            return ref field;
+        }
+    }
+
+    int this[int x, long y, int z, int opt = 1]
+    {
+        set
+        {
+            Console.WriteLine($""this[x: {x}, y: {y}, z: {z}].set({value})"");
+        }
+    }
+
+    int M(int i)
+    {
+        Console.WriteLine($""M({i})"");
+        return 0;
+    }
+
+    void Test()
+    {
+        (this[z: M(1), x: M(2), y: 10], this[z: M(3), x: M(4), y: 10L]) = this;
+        Console.WriteLine($""field: {field}"");
+    }
+
+    static void Main()
+    {
+        new C().Test();
+    }
+
+    void Deconstruct(out int a, out int b)
+    {
+        Console.WriteLine(nameof(Deconstruct));
+        a = 1;
+        b = 2;
+    }
+}
+";
+
+            var expectedOutput =
+@"M(1)
+M(2)
+this[x: 0, y: 10, z: 0].get
+M(3)
+M(4)
+Deconstruct
+this[x: 0, y: 10, z: 0].set(2)
+field: 1
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: expectedOutput, additionalRefs: new[] { ValueTupleRef, SystemRuntimeFacadeRef });
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(18554, "https://github.com/dotnet/roslyn/issues/18554")]
+        public void AssigningTupleIntoIndexers()
```

Comment:
>AssigningTupleIntoIndexers [](start = 20, length = 26)

It feels like we should also add tests for scenarios when the ```value``` parameter of the setter is optional. If we cannot declare such setter in source, this can be done with IL.